ff-fft/src/domain/mod.rs

```diff
@@ -81,20 +88,62 @@ impl<F: PrimeField> EvaluationDomain<F> {
 
     /// Construct a domain that is large enough for evaluations of a polynomial
     /// having `num_coeffs` coefficients.
+    ///
+    /// If the field specifies a small subgroup for a mixed-radix FFT and
+    /// the radix-2 FFT cannot be constructed, this method tries
+    /// constructing a mixed-radix FFT instead.
     pub fn new(num_coeffs: usize) -> Option<Self> {
-        // Compute the size of our evaluation domain
-        let size = num_coeffs.next_power_of_two() as u64;
-        let log_size_of_group = size.trailing_zeros();
+        let domain = Self::construct(num_coeffs);
+        if domain.is_some() {
+            return domain;
+        }
 
-        if log_size_of_group >= F::Params::TWO_ADICITY {
-            return None;
+        if F::Params::SMALL_SUBGROUP_BASE.is_some() {
+            let s = best_mixed_domain_size::<F>(num_coeffs);
+            return Self::construct(s);
         }
 
-        // Compute the generator for the multiplicative subgroup.
-        // It should be 2^(log_size_of_group) root of unity.
-        let mut group_gen = F::root_of_unity();
-        for _ in log_size_of_group..F::Params::TWO_ADICITY {
-            group_gen.square_in_place();
+        None
+    }
+
+    /// Construct a domain that is large enough for evaluations of a polynomial
+    /// having `num_coeffs` coefficients.
+    pub fn construct(num_coeffs: usize) -> Option<Self> {
+        // Compute the size of our evaluation domain
+        let size;
+        let log_size_of_group;
+        let group_gen;
+
+        if let Some(small_subgroup_base) = F::Params::SMALL_SUBGROUP_BASE {
+            let q = small_subgroup_base as usize;
+            let q_adicity = k_adicity(q, num_coeffs);
+            let q_part = q.pow(q_adicity);
+
+            let two_adicity = k_adicity(2, num_coeffs);
+            let two_part = 1 << two_adicity;
+
+            size = num_coeffs as u64;
+            log_size_of_group = two_adicity;
+
+            if num_coeffs != q_part * two_part {
+                return None;
+            }
+
+            group_gen = F::get_root_of_unity(num_coeffs)?;
+        } else {
+            // Compute the size of our evaluation domain
+            size = num_coeffs.next_power_of_two() as u64;
+            log_size_of_group = size.trailing_zeros();
+
+            // TODO: Check > vs. >= here.
+            // libfqfft uses > https://github.com/scipr-lab/libfqfft/blob/e0183b2cef7d4c5deb21a6eaf3fe3b586d738fe0/libfqfft/evaluation_domain/domains/basic_radix2_domain.tcc#L33
+            if log_size_of_group > F::Params::TWO_ADICITY {
```

Comment:
This should be re-checked. Zexe's implementation previously did not allow `log_size_of_group` to be equal to the 2-adicity. libfqfft does allow that, though.

Reply:
As far as I can see, there is no reason preventing the FFT code to run with the 2^s root of unity, which is a normal root of unity after all. The code for `get_root_of_unity` also handles this case fine.